Code specialization

AndrasKovacs opened the pull request:
Decrease `-foptimise-heavily` build time by 15-17% by adding a bunch of `SPECIALIZE` pragmas. Also add a bunch of `INLINE` and `INLINABLE` pragmas. `-foptimise-heavily` runtime performance is unchanged within margin of error. 

Pragmas are placed based on the following information:
- Module build times displayed with https://github.com/codedownio/time-ghc-modules and `-ddump-timings -ddump-to-file` GHC options. 
- Core printed with `-ddump-simpl -dsuppress-all -dno-suppress-type-signatures -ddump-to-file`.
- Failed specializations printed with `-Wall-missed-specializations` when we build without `-foptimise-heavily`. 

Only the worst offending and most visible instances of code duplication are addressed here; there is definitely a large amount that's still there.

andreasabel commented on src/full/Agda/Syntax/Translation/InternalToAbstract.hs:
```diff
@@ -408,14 +417,15 @@ instance Reify Literal where
 
   reifyWhen = reifyWhenE
-  reify v = reifyTerm True v
+  reify v = reifyTerm True v; {-# inline reify #-}
```
INLINE

andreasabel commented on src/full/Agda/Interaction/BasicOps.hs:
```diff
@@ -426,80 +425,82 @@ reifyElimToExpr = \case
+    OfType <$> reify (MetaV m []) <*> reify t
+  reify (CheckType t) = JustType <$> reify t
+  reify (UsableAtModality _ _ mod t) = UsableAtMod mod <$> reify t
+  {-# SPECIALIZE reify :: Constraint -> TCM (ReifiesTo Constraint) #-}
```
There is a large diff just fixing some indentation.  Could whitespace changes be put into a separate commit please? (can then be ignored in a git blame).

andreasabel commented on src/full/Agda/TypeChecking/Pretty.hs:
```diff
@@ -161,36 +166,55 @@ prettyTCMCtx :: (PrettyTCM a, MonadPretty m) => Precedence -> a -> m Doc
+instance PrettyTCM Nat                        where prettyTCM = pretty
+instance PrettyTCM ProblemId                  where prettyTCM = pretty
+instance PrettyTCM Range                      where prettyTCM = pretty
+instance PrettyTCM CheckpointId               where prettyTCM = pretty
```
Please put this whitespace change into a separate commit.

andreasabel commented on src/full/Agda/Syntax/Position.hs:
```diff
@@ -735,7 +737,7 @@ x `withRangeOf` y = setRange (getRange y) x
 --   first list is placed first.
 interleaveRanges :: (HasRange a) => [a] -> [a] -> ([a], [(a,a)])
-interleaveRanges as bs = runWriter$ go as bs
+interleaveRanges as bs = runWriter $ go as bs
```
whitespace

andreasabel commented on src/full/Agda/Syntax/Translation/AbstractToConcrete.hs:
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE CPP          #-}
+{-# LANGUAGE CPP #-}
```
whitespace

andreasabel commented on src/full/Agda/Interaction/BasicOps.hs:
```diff
@@ -1,5 +1,4 @@
 {-# LANGUAGE NondecreasingIndentation #-}
-
```
whitespace

andreasabel commented on src/full/Agda/TypeChecking/CheckInternal.hs:
```diff
@@ -219,8 +226,8 @@ checkArgInfo action ai ai' = do
-checkHiding    :: (MonadCheckInternal m) => Hiding -> Hiding -> m ()
-checkHiding    h h' = unless (sameHiding h h') $ typeError $ HidingMismatch h h'
+checkHiding :: (MonadCheckInternal m) => Hiding -> Hiding -> m ()
+checkHiding h h' = unless (sameHiding h h') $ typeError $ HidingMismatch h h'
```
whitespace

andreasabel reviewed (commented):
Thanks, this is great!

One request for commit hygiene:  Could you please gather the whitespace changes into a whitespace-only commit?  This is good when coming back to these commits in the future, e.g. via blame or bisect.

AndrasKovacs commented:
I moved the whitespace edits to a commit.

AndrasKovacs committed:
Add SPECIALIZE, INLINE, INLINABLE pragmas

AndrasKovacs committed:
Whitespace changes

Newlines, indentation, alignment, putting the "::" right after top-level names
for searchability.

andreasabel commented on src/full/Agda/TypeChecking/Monad/MetaVars.hs:
```diff
@@ -774,15 +775,13 @@ solveAwakeConstraints' = solveSomeAwakeConstraints (const True)
-  freeze ::
-    Monad m =>
-    MetaId -> StateT LocalMetaStore (WriterT (Set MetaId) m) ()
+  freeze :: MetaId -> StateT LocalMetaStore (WriterT (Set MetaId) m) ()
```
Borderline whitespace change :-D (not really just whitespace, but ok).

andreasabel commented on src/full/Agda/TypeChecking/Monad/Base.hs:
```diff
@@ -3820,9 +3828,7 @@ eQuantity f e =
-eHardCompileTimeMode f e =
-  f (envHardCompileTimeMode e) <&>
-  \x -> e { envHardCompileTimeMode = x }
+eHardCompileTimeMode f e = f (envHardCompileTimeMode e) <&> \x -> e { envHardCompileTimeMode = x }
```
Looks like a whitespace change.  Oh well.

andreasabel reviewed (approved):
Thanks for the update!

The sorting into ws/non-ws is now 99% correct, let's ignore the 2 outliers.

andreasabel commented on src/full/Agda/TypeChecking/Pretty.hs:
```diff
@@ -175,22 +180,37 @@ instance PrettyTCM CheckpointId where prettyTCM = pretty
+{-# SPECIALIZE prettyTCM :: Nat                -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: ProblemId          -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: Range              -> TCM Doc #-}
+{-# SPECIALIZE prettyTCM :: CheckpointId       -> TCM Doc #-}
```
Why did these functions get SPECIALIZE rather then INLINE pragmas?
(Converse question below.)

andreasabel commented on src/full/Agda/TypeChecking/Pretty.hs:
```diff
@@ -274,48 +319,54 @@ instance PrettyTCM Elim where
 
 instance PrettyTCM A.Expr where
-  prettyTCM = prettyA
+  prettyTCM = prettyA; {-# INLINE prettyTCM #-}
```
Why did these functions get INLINE pragmas rather than SPECIALIZE ones?

andreasabel replied:
@AndrasKovacs answered that SPECIALIZE for all would be safer.
